lib/rex/proto/ldap/server.rb

```diff
@@ -149,104 +155,177 @@ def dispatch_request(cli, data)
-              code.to_ber_enumerated,
-              dn.to_ber,
-              msg.to_ber
-            ].to_ber_appsequence(tag)
+            tag_sequence.to_ber_appsequence(tag)
           ].to_ber_sequence
         end
 
         #
-        # Search provided ldif data for query information
+        # Search provided ldif data for query information. If no `ldif` was provided a random search result will be generated.
         #
         # @param filter [Net::LDAP::Filter] LDAP query filter
         # @param attrflt [Array, Symbol] LDAP attribute filter
         #
         # @return [Array] Query matches
-        def search_ldif(filter, msgid, attrflt = :all)
-          return [] if @ldif.nil? || @ldif.empty?
-
-          ldif.map do |dn, entry|
-            next unless filter.match(entry)
 
+        def search_result(filter, msgid, attrflt = :all)
+          if @ldif.nil? || @ldif.empty?
             attrs = []
-            entry.each do |k, v|
-              if attrflt == :all || attrflt.include?(k.downcase)
-                attrvals = v.map(&:to_ber).to_ber_set
-                attrs << [k.to_ber, attrvals].to_ber_sequence
+            if attrflt.is_a?(Array)
+              attrflt.each do |at|
+                attrval = [Rex::Text.rand_text_alphanumeric(10)].map(&:to_ber).to_ber_set
+                attrs << [at.to_ber, attrval].to_ber_sequence
               end
+              dn = "dc=#{Rex::Text.rand_text_alphanumeric(10)},dc=#{Rex::Text.rand_text_alpha(4)}"
+              appseq = [
+                dn.to_ber,
+                attrs.to_ber_sequence
+              ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
+              [msgid.to_ber, appseq].to_ber_sequence
             end
-            appseq = [
-              dn.to_ber,
-              attrs.to_ber_sequence
-            ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
-            [msgid.to_ber, appseq].to_ber_sequence
-          end.compact
+          else
+            ldif.map do |bind_dn, entry|
+              next unless filter.match(entry)
+
+              attrs = []
+              entry.each do |k, v|
+                if attrflt == :all || attrflt.include?(k.downcase)
+                  attrvals = v.map(&:to_ber).to_ber_set
+                  attrs << [k.to_ber, attrvals].to_ber_sequence
+                end
+              end
+              appseq = [
+                bind_dn.to_ber,
+                attrs.to_ber_sequence
+              ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
+              [msgid.to_ber, appseq].to_ber_sequence
+            end.compact.join
+          end
+        end
+
+        #
+        # Sets the tasks to be performed after processing of pdu object
+        #
+        # @param proc [Proc] block of code to execute
+        #
+        # @return pdu_process [Proc] steps to be executed
+        def processed_pdu_handler(pdu_type, &proc)
+          @pdu_process = []
+          @pdu_process[pdu_type] = proc if block_given?
```

Comment:
No required at this time, consider adding registered processors to a list for the type allowing multiple actions based on a single type.

```suggestion
          if block_given?
            @pdu_process[pdu_type] = [] unless @pdu_process[pdu_type]
            @pdu_process[pdu_type] << proc
          end
```

Note that if adjusted to accept a set of processors, there also needs to be a way to clear all registered processors or remove a specific processor when a module exits expecting to leave the service running so some sort of identifier for the proc to be removed may be needed.

Reply:
We could add something unique to the modules that can be used to identify the processors specific to that module if we want to add multiple processors

Reply:
The caller to `processed_pdu_handler` is an instance of a module that is unique during runtime so the mixin layer could pass a reference to `self` that the service could hold and use for comparison when handlers are removed.

The complexity of holding references though requires careful validation testing, holding a reference in this way can cause memory leaks if not released properly resulting in the object referenced never being considered ready for garbage collection.